pkg/client/unversioned/clientcmd/merged_client_builder.go

```diff
@@ -112,8 +112,11 @@ func (config *DeferredLoadingClientConfig) ClientConfig() (*restclient.Config, e
 		//   "empty due to defaults"
 		// TODO: this shouldn't be a global - the client config rules should be
 		//   handling this.
-		defaultConfig, err := DefaultClientConfig.ClientConfig()
-		if err == nil && !reflect.DeepEqual(mergedConfig, defaultConfig) {
+		defaultConfig, defErr := DefaultClientConfig.ClientConfig()
+		if IsConfigurationInvalid(defErr) && !IsEmptyConfig(err) {
```

Comment:
@smarterclayton I agree with all the cases here except:

> The cases that should fall through to in-cluster are

> 1. base config is valid and non-default

Why should this fall through to the in-cluster config? This is a problem, for example, for federation. The config is of federation API server and non-default. It cannot just return in-cluster config.

I think this is an unfortunate consequence of an explicit lack of contract for `DefaultClientConfig`. In my view, constructed from grepping through Kubernetes source code, `DefaultClientConfig` is essentially a const. I don't see who/where it is modified. So it can never be non-empty or valid with the way it is now. But you probably have a different view where something is modifying the `DefaultClientConfig`.

But anyway, this is what is happening today. for this case. This isn't a problem for federation right now, because we never provide a valid default.

I think we should get this merged to unblock federation and the release. and debate about the cases separately.